android/app/src/main/java/com/microsoft/codepush/common/core/CodePushBaseCore.java

```diff
@@ -0,0 +1,1043 @@
+     * @throws CodePushPlatformUtilsException if error occurred during usage of {@link CodePushPlatformUtils}.
+     * @throws CodePushRollbackException      if error occurred during rolling back of package.
+     */
+    @SuppressWarnings("WeakerAccess")
+    protected void initializeUpdateAfterRestart() throws CodePushGetPackageException, CodePushRollbackException, CodePushPlatformUtilsException {
+
+        /* Reset the state which indicates that
+         * the app was just freshly updated. */
+        mState.mDidUpdate = false;
+        CodePushPendingUpdate pendingUpdate = mManagers.mSettingsManager.getPendingUpdate();
+        if (pendingUpdate != null) {
+            CodePushLocalPackage packageMetadata = mManagers.mUpdateManager.getCurrentPackage();
+            if (packageMetadata == null || !mUtilities.mPlatformUtils.isPackageLatest(packageMetadata, mAppVersion) &&
+                    !mAppVersion.equals(packageMetadata.getAppVersion())) {
+                AppCenterLog.info(LOG_TAG, "Skipping initializeUpdateAfterRestart(), binary version is newer.");
+                return;
+            }
+            boolean updateIsLoading = pendingUpdate.isPendingUpdateLoading();
+            if (updateIsLoading) {
+
+                /* Pending update was initialized, but notifyApplicationReady was not called.
+                 * Therefore, deduce that it is a broken update and rollback. */
+                AppCenterLog.info(LOG_TAG, "Update did not finish loading the last time, rolling back to a previous version.");
+                mState.mNeedToReportRollback = true;
+                rollbackPackage();
+            } else {
+
+                /* There is in fact a new update running for the first
+                 * time, so update the local state to ensure the client knows. */
+                mState.mDidUpdate = true;
+
+                /* Mark that we tried to initialize the new update, so that if it crashes,
+                 * we will know that we need to rollback when the app next starts. */
+                mManagers.mSettingsManager.savePendingUpdate(pendingUpdate);
+            }
+        }
+    }
+
+    /**
+     * Rolls back package.
+     *
+     * @throws CodePushGetPackageException if error occurred during getting current update.
+     * @throws CodePushRollbackException   if error occurred during rolling back of package.
+     */
+    private void rollbackPackage() throws CodePushGetPackageException, CodePushRollbackException {
+        CodePushLocalPackage failedPackage = mManagers.mUpdateManager.getCurrentPackage();
+        mManagers.mSettingsManager.saveFailedUpdate(failedPackage);
+        mManagers.mUpdateManager.rollbackPackage();
+        mManagers.mSettingsManager.removePendingUpdate();
+    }
+
+    /**
+     * Clears any saved updates on device.
+     *
+     * @throws IOException if i/o error occurred while accessing the file system.
```

Comment:
Please use default javadocs for exception where it's possible and reasonable (see pinned items in Slack).

Reply:
@annakocheshkova got it.